Sources/SIL/Parser.swift

```diff
@@ -48,6 +48,17 @@ class Parser {
         return chars.suffix(from: cursor).starts(with: Array(query))
     }
 
+    // Applies the function, but restores the cursor from before the call if it returns nil.
+    func tryParse<T>(_ f: () throws -> T?) rethrows -> T? {
```

Comment:
Also I think it would be useful to move this one next to other `parseFoo` combinators.